### PR TITLE
ledger: generic kv trackers backend implementation

### PR DESCRIPTION
**Pull request by icorderi**
## Summary

We are breaking down #5212 in several smaller PRs to make it more digestible. 

This is PRs "numero 3" on the KV epic.

Content:
- A generic KV implementation with behavioral parity with the SQLite implementation 
- A testsuite to run tests across multiple db backends 

## Test Plan

Adding cross-backend db testing.

### CODE REVIEW

File: ledger/store/trackerdb/generickv/accounts_ext_reader.go [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+func (r *accountsReader) TotalKVs(ctx context.Context) (total uint64, err error) {
+	// TODO: catchpoint
+	return
+}
+
+// TODO: this replicates some functionality from LookupOnlineHistory, implemented for onlineAccountsReader
+func (r *accountsReader) LookupOnlineAccountDataByAddress(addr basics.Address) (ref trackerdb.OnlineAccountRef, data []byte, err error) {
+	low := onlineAccountOnlyPartialKey(addr)
+	high := onlineAccountOnlyPartialKey(addr)
+	high[len(high)-1]++
```
Comment (algorandskiy) [inline ledger/store/trackerdb/generickv/accounts_ext_reader.go:128]: nit: maybe have a `exclusive=true` flag or something. Plus, some methods use `+ "."`, others `high[len(high)-1]++`, we need some consistency here

File: ledger/store/trackerdb/testsuite/onlineaccounts_kv_test.go [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+	require.NoError(t, err)
+	require.Equal(t, basics.MicroAlgos{Raw: uint64(42)}, totals)
+
+	// lookup not found
+	_, err = oaor.LookupOnlineTotalsHistory(basics.Round(121))
+	require.Error(t, err)
+	require.Equal(t, trackerdb.ErrNotFound, err)
+}
+
+func CustomTestOnlineAccountsDelete(t *customT) {
```
Comment (algorandskiy) [inline ledger/store/trackerdb/testsuite/onlineaccounts_kv_test.go:358]: there is a bigger test `TestOnlineAccountsDeletion` and I do think the new implementation have to at least pass it but ideally extend it to 3+ accounts and more rounds.

File: ledger/store/trackerdb/testsuite/utils_test.go [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+// list of tests to be run on each KV DB implementation
+var genericTests []genericTestEntry
+
+// registerTest registers the given test with the suite
+func registerTest(name string, f func(*customT)) {
+	genericTests = append(genericTests, genericTestEntry{name, f})
+}
+
+// runGenericTestsWithDB runs a generic set of tests on the given database
+func runGenericTestsWithDB(t *testing.T, dbFactory func(config.ConsensusParams) (db dbForTests)) {
```
Comment (algorandskiy) [inline ledger/store/trackerdb/testsuite/utils_test.go:67]: what about wrapping this and running sql and kv, and then comparing data in both? like to check other side effects that are out of test asserts?

**Review by algorandskiy — COMMENTED** [review]
Looks good but there are couple primary concerns:
1. new `OnlineAccountsDelete` needs more test scenarios. The original SQL code was written for continuous address sequence like AAAABBBCCCC and not ABCABAACC.
2. keys creation has unneeded allocation overhead, need to fix.

File: ledger/store/trackerdb/testsuite/utils_test.go (hunk 3 above)
Comment (icorderi) [reply ledger/store/trackerdb/testsuite/utils_test.go:67]: The `dualdriver` does exactly that, it is also plugged in into this tests. 
We are still missing the PR that brings that in, after we have the `pebbledriver` PR.

File: ledger/store/trackerdb/generickv/accounts_ext_reader.go [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+	}
+	return
+}
+
+func (r *accountsReader) AccountsOnlineRoundParams() (onlineRoundParamsData []ledgercore.OnlineRoundParamsData, endRound basics.Round, err error) {
+	// The SQL at the time of writing:
+	//
+	// SELECT rnd, data FROM onlineroundparamstail ORDER BY rnd ASC
+
+	start := []byte(kvOnlineAccountRoundParams + "-")
```
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/store/trackerdb/generickv/accounts_ext_reader.go:257]: No, it's not a regex. 
"." just comes some ascii values after "-". 

I just a way to define the prefix range, not 100% byte correct if we had different separators.

File: ledger/store/trackerdb/generickv/accounts_ext_writer.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+		rndOffset := len(kvPrefixOnlineAccountBalance) + 1
+		u64Rnd := binary.BigEndian.Uint64(key[rndOffset : rndOffset+8])
+		round := basics.Round(u64Rnd)
+
+		// get the offset where the address starts
+		addrOffset := len(kvPrefixOnlineAccountBalance) + 1 + 8 + 1 + 8 + 1
+		var addr basics.Address
+		copy(addr[:], key[addrOffset:addrOffset+32])
+
+		if addr != prevAddr {
```
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/store/trackerdb/generickv/accounts_ext_writer.go:165]: I'll double check, the tests for methods that return multiple values do an ordering check. 
The dualengine (missing from this PR) which runs pebble+sqlite and compares the outputs also checks ordering. 

I could have missed the explicit ordering check on this one.

File: ledger/store/trackerdb/generickv/accounts_ext_reader.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/store/trackerdb/generickv/accounts_ext_reader.go:257]: yeah, I figured out after (see the last comment). It is confusing somewhere we set it to ".", somewhere just increment '-'.

File: ledger/store/trackerdb/generickv/accounts_reader.go [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+	data.Addr = addr
+
+	// read the current db round
+	data.Round, err = r.AccountsRound()
+	if err != nil {
+		return
+	}
+
+	value, closer, err := r.kvr.Get(accountKey(addr))
+	if err == trackerdb.ErrNotFound {
```
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/store/trackerdb/generickv/accounts_reader.go:85]: Fair point, lets just assume its a bug on the db driver implementation not to return the generic `trackerdb.ErrNotFound`.

**Review by AlgoAxel — APPROVED** [review]
As I have used/been exposed to this code previously, I:
* Read over existing comments
* Skimmed over all
* Read closely sections I was not already familiar with (stateproof)

As I already use this code (in a slightly different format), I am comfortable approving.

**Comment by algorandskiy** [comment]
Going to merge after the 3.16.3 cut and have follow up ticket for tasks.

File: ledger/store/trackerdb/generickv/accounts_ext_reader.go [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+	// TODO: catchpoint
+	return
+}
+
+func (r *accountsReader) LookupAccountAddressFromAddressID(ctx context.Context, ref trackerdb.AccountRef) (address basics.Address, err error) {
+	// TODO: catchpoint
+	return
+}
+
+func (r *accountsReader) LookupAccountRowID(addr basics.Address) (ref trackerdb.AccountRef, err error) {
```
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/store/trackerdb/generickv/accounts_ext_reader.go:77]: correct, this is an artifact from when the `normBalance` was part of the key.

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**